libraries/EEPROM/src/EEPROM.cpp

```diff
@@ -66,31 +63,59 @@ EEPROMClass::~EEPROMClass() {
 }
 
 bool EEPROMClass::begin(size_t size) {
-  if (size <= 0) {
-    return false;
+  if (size <= 0) return false;
+
+  esp_err_t res = nvs_open(_name, NVS_READWRITE, &_handle);
+  if (res != ESP_OK) {
+      log_e("Unable to open NVS namespace: %d", res);
+      return false;
+  }
+
+  size_t key_size = 0;
+  res = nvs_get_blob(_handle, _name, NULL, &key_size);
+  if(res != ESP_OK && res != ESP_ERR_NVS_NOT_FOUND) {
+      log_e("Unable to read NVS key: %d", res);
+      return false;
   }
-  if (size > SPI_FLASH_SEC_SIZE) {
-    size = SPI_FLASH_SEC_SIZE;
+  if (size < key_size) { // truncate
+      log_w("truncating EEPROM from %d to %d", key_size, size);
+      uint8_t* trunc_key = (uint8_t*) malloc(size);
+      nvs_get_blob(_handle, _name, trunc_key, &size);
+      nvs_set_blob(_handle, _name, trunc_key, size);
+      nvs_commit(_handle);
+      free(trunc_key);
   }
-  //  _mypart = esp_partition_find_first(ESP_PARTITION_TYPE_DATA,ESP_PARTITION_SUBTYPE_ANY, EEPROM_FLASH_PARTITION_NAME);
-  _mypart = esp_partition_find_first(ESP_PARTITION_TYPE_DATA, ESP_PARTITION_SUBTYPE_ANY, _name);
-  if (_mypart == NULL) {
-    return false;
+  if (size > key_size) { // expand or new
+      size_t expand_size = size - key_size;
+      uint8_t* expand_key = (uint8_t*) malloc(expand_size);
+      if(nvs_set_blob(_handle, "expand", expand_key, expand_size)) {
+        log_e("Not enough space to expand EEPROM from %d to %d", key_size, size);
+        free(expand_key);
+        return false;
+      }
+      free(expand_key);
+      nvs_erase_key(_handle, "expand");
+      uint8_t* hold_data = (uint8_t*) malloc(size);
+      memset(hold_data, 0, size);
+      if(key_size) {
+        log_i("Expanding EEPROM from %d to %d", key_size, size);
+        nvs_get_blob(_handle, _name, hold_data, &key_size);
+        nvs_erase_key(_handle, _name);
+      }
+      nvs_commit(_handle);
+      nvs_set_blob(_handle, _name, hold_data, size);
+      free(hold_data);
+      nvs_commit(_handle);
```

Comment:
this whole block of code needs some comments to explain exactly what is going on :) I got lost with all those reads, writes and erases. What data is where... you need 2 blobs to hold the default 4K EEPROM, but reading through your code I got lost.

Reply:
There is no way to expand a key, so in order to do so, it tests for free space, then holds the data while it deletes the key.  If it is a new key, it just fills with zeros.  I think changing the variable name to key_data will make it clearer.